### PR TITLE
add weighted queries for hybrid search

### PR DESCRIPTION
**Pull request by danyilq**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Introduces `queryTensor` and `queryLexical` hybrid parameters. It also introduces the support of weighted tensor queries.

* **What is the current behavior?** (You can also link to an open issue here)
Currently weighted queries can not be used in tensor search.

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)

* **Related Python client changes** (link commit/PR here)

* **Related documentation changes** (link commit/PR here)
https://github.com/marqo-ai/marqodocs/pull/422

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py [hunk 1]
```diff
@@ -453,3 +454,71 @@ def test_searchable_attributes(self):
...
+            },
+            {
+                "title": "Just Your Average Doc",
+                "content": "this is a solid doc",
+                "_id": "2"
+            }
+        ]
+        self.client.index(self.text_index_name).add_documents(docs, tensor_fields=["title", "content"])
+        # Hybrid search with no query or context should raise an error
+        with self.assertRaises(MarqoWebError):
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py]: Use subtests or separate test methods

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py [hunk 2]
```diff
@@ -453,3 +454,71 @@ def test_searchable_attributes(self):
...
+            )
+
+        # Hybrid search with no query or context should raise an error
+        with self.assertRaises(MarqoWebError):
+            self.client.index(self.text_index_name).search(
+                search_method="HYBRID",
+                hybrid_parameters={}
+            )
+
+        # Hybrid search with query and queryTensor/queryLexical should raise an error
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py]: Would be good to assert the message too, otherwise you could be getting the wrong error (in which case the test should fail not pass)

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py [hunk 3]
```diff
@@ -453,3 +454,71 @@ def test_searchable_attributes(self):
...
+            }
+        ]
+        self.client.index(self.text_index_name).add_documents(docs, tensor_fields=["title", "content"])
+        # Hybrid search with no query or context should raise an error
+        with self.assertRaises(MarqoWebError):
+            self.client.index(self.text_index_name).search(
+                search_method="HYBRID"
+            )
+
+        # Hybrid search with no query or context should raise an error
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py]: With subtests this comment can be test description

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py [hunk 4]
```diff
@@ -453,3 +454,71 @@ def test_searchable_attributes(self):
                                                                       searchable_attributes=["title"])
         self.assertEqual(len(search_res["hits"]), 1)
         self.assertEqual(search_res["hits"][0]["_id"], "1")
+
+    def test_hybrid_search_validations(self):
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py]: These tests don't seem specific to unstructured?

File: tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py [hunk 5]
```diff
@@ -2714,4 +2714,100 @@ def test_correct_error_is_raised_if_rerank_depth_is_provided_on_old_marqo_versio
...
+                )
+
+                # Reference results
+                tensor_res = tensor_search.search(
+                    config=self.config,
+                    index_name=self.structured_text_index_score_modifiers.name,
+                    search_method="HYBRID",
+                    text=None,
+                    hybrid_parameters=HybridParameters(
+                        queryTensor={
```
Comment (farshidz) [inline tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py]: We need more coverage, including: 
Lexica-lexical
Tensor-tensor
Tensor-lexical

Also for the tensor dict, we need 
* One query (covered)
* 3 queries (avg case)
* Empty dict (edge case)
* String tensor query (i.e. not dict)

Other tests you should have are:
* Tensor/tensor with string tensor query or dict with one entry matches standard tensor search
* Same as above but for lexical/lexical

We're also not really proving the weighted average is done right. Perhaps you could do some monkey patching to get the averaged vector and have an assert on that.

Furthermore, API test coverage is insufficient. We need at least one test for each mode (disjunction, lexical/lexical etc). You don't have to test different tensor query sizes there so integ tests are always more thorough

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 from marqo import enums
 from marqo.client import Client
 from marqo.enums import SearchMethods
+from marqo.errors import MarqoWebError
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py:9]: New API must also be covered in backwards compatibility tests

File: src/marqo/core/search/hybrid_search.py [hunk 7]
```diff
@@ -128,19 +128,29 @@ def search(
...
         text_query_prefix = marqo_index.model.get_text_query_prefix(text_query_prefix)
+        # split queries into lexical and tensor
+        if query is None:
+            tensor_query = hybrid_parameters.queryTensor
+            lexical_query = hybrid_parameters.queryLexical
+        elif isinstance(query, CustomVectorQuery):
+            tensor_query = query
+            lexical_query = None
+        else:
+            tensor_query = query
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py:155]: Are you relying on external validation here? Because you could have query and tensor query and lexical query all provided for example, then you'll just take query and ignore the rest. Even if there's higher level validation, this method should validate its input independently as a sanity check and to make the code easier to understand and reason about without assumptions about the caller.

File: src/marqo/core/search/hybrid_search.py [hunk 8]
```diff
@@ -155,8 +165,8 @@ def search(
             query_text_search = None
 
         else:  # string or dict query
-            query_text_vectorise = query
-            query_text_search = query
+            query_text_vectorise = tensor_query if tensor_query else ""
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py]: I suspect None and "" will not behave the same here. Have you verified that? e.g. empty string almost certainly will create a vector. A disjunction query with just a lexical query should instead be an error and similarly for other modes.

File: src/marqo/tensor_search/models/api_models.py [hunk 9]
```diff
@@ -100,16 +100,37 @@ def validate_query_and_context(cls, values):
         search_method = values.get('searchMethod')
         query = values.get('q')
         context = values.get('context')
+        hybrid_parameters = values.get('hybridParameters')
```
Comment (farshidz) [inline src/marqo/tensor_search/models/api_models.py:103]: This code needs to be unit-tested

File: src/marqo/tensor_search/models/api_models.py [hunk 10]
```diff
@@ -100,16 +100,37 @@ def validate_query_and_context(cls, values):
...
+                    )
+                ):
+                    raise ValueError(
+                        f"One of Query(q), context, hybridParameters.queryTensor, or "
+                        f"hybridParameters.contextTensor is required for {search_method} search but all are missing"
+                    )
         else:
-            raise ValueError(f"Invalid search method {search_method}")
+            if search_method == SearchMethod.HYBRID:
+                if cls.__name__ == 'SearchQuery': # This check is only relevant for Initial SearchQuery, not BulkSearchQuery
```
Comment (farshidz) [inline src/marqo/tensor_search/models/api_models.py]: I don't quite understand this class check here. Are you making sure subclasses don't run this?

File: src/marqo/tensor_search/validation.py [hunk 11]
```diff
@@ -43,8 +43,11 @@ def validate_query(q: Optional[Union[dict, str, CustomVector]], search_method: U
 
         if search_method.upper() != SearchMethod.TENSOR:
             raise InvalidArgError(
-                'Multi-query search is currently only supported for search_method="TENSOR" '
-                f"\nReceived search_method `{search_method}`. {usage_ref}")
+                'Multi-query search is only supported for search_method="TENSOR" or "HYBRID".'
+                f'\nReceived invalid search_method: `{search_method}`.'
+                '\nNote: For HYBRID search, use `hybrid_parameters.queryTensor` instead of `q` for multi-query input.'
```
Comment (farshidz) [inline src/marqo/tensor_search/validation.py:48]: This is a nice touch

**Review by farshidz — CHANGES_REQUESTED** [review]

File: tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py (hunk 6 above)
Comment (farshidz) [reply tests/api_tests/v1/tests/api_tests/unstructured_index/test_search.py:9]: Agreed will be a follow-up PR

File: src/marqo/core/search/hybrid_search.py [hunk 12]
```diff
@@ -124,6 +124,12 @@ def search(
                 f"`searchableAttributesLexical`. Please set these attributes to None."
             )
 
+        if query and hybrid_parameters.queryLexical and hybrid_parameters.queryTensor:
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py]: I think should be 
```
query and (hybrid_parameters.queryLexical or hybrid_parameters.queryTensor)
```

File: src/marqo/tensor_search/models/api_models.py (hunk 10 above)
Comment (danyilq) [reply src/marqo/tensor_search/models/api_models.py]: improved comment so it makes more sense

File: src/marqo/core/search/hybrid_search.py (hunk 12 above)
Comment (farshidz) [reply src/marqo/core/search/hybrid_search.py]: Also you want to distinguish between None and empty string. So query defined and lexicalQuery empty string is an error, but this won't eval to True. In general `if x` tends to create bugs in Python, as opposed to `if x is None` etc

File: src/marqo/core/search/hybrid_search.py [hunk 13]
```diff
@@ -163,8 +169,8 @@ def search(
             query_text_search = None
 
         else:  # string or dict query
-            query_text_vectorise = tensor_query if tensor_query else ""
-            query_text_search = lexical_query if lexical_query else ""
+            query_text_vectorise = tensor_query
+            query_text_search = lexical_query
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py:192]: Have you tested passing None lexical or tensor to this? I suspect it'll error out. I think you need validation (and tests) for this e.g., if disjunction or lexical/tensor or tensor/lexical must provide both (or just q), if lexical/lexical or tensor/tensor must provide exactly one etc

File: src/marqo/tensor_search/models/api_models.py [hunk 14]
```diff
@@ -120,11 +120,14 @@ def validate_query_and_context(cls, values):
...
+                        f"hybridParameters.queryTensor is required for {search_method} search but all are missing"
                     )
         else:
             if search_method == SearchMethod.HYBRID:
-                if cls.__name__ == 'SearchQuery': # This check is only relevant for Initial SearchQuery, not BulkSearchQuery
+                if cls.__name__ == 'SearchQuery':
+                    # This check is needed because BulkSearchQuery inherits SearchQuery and because of the way we set
+                    # query for it, it causes this check to fail since we previously provided queryTensor/queryLexical
+                    # parameters
                     if hybrid_parameters and (hybrid_parameters.queryTensor or hybrid_parameters.queryLexical):
```
Comment (farshidz) [inline src/marqo/tensor_search/models/api_models.py]: I think these checks will fail to work as expected for empty string because e.g., `hybrid_parameters.queryTensor` evals to False for empty string but it's technically a value still. Unit tests should cover these edge cases

File: src/marqo/tensor_search/models/api_models.py [hunk 15]
```diff
@@ -120,11 +120,14 @@ def validate_query_and_context(cls, values):
                 ):
                     raise ValueError(
                         f"One of Query(q), context, hybridParameters.queryTensor, or "
-                        f"hybridParameters.contextTensor is required for {search_method} search but all are missing"
+                        f"hybridParameters.queryTensor is required for {search_method} search but all are missing"
                     )
         else:
             if search_method == SearchMethod.HYBRID:
-                if cls.__name__ == 'SearchQuery': # This check is only relevant for Initial SearchQuery, not BulkSearchQuery
+                if cls.__name__ == 'SearchQuery':
```
Comment (farshidz) [inline src/marqo/tensor_search/models/api_models.py]: Why not `if cls is SearchQuery:`?

**Review by farshidz — DISMISSED** [review]

File: src/marqo/tensor_search/models/api_models.py [hunk 16]
```diff
@@ -124,11 +124,13 @@ def validate_query_and_context(cls, values):
...
         else:
             if search_method == SearchMethod.HYBRID:
-                if cls.__name__ == 'SearchQuery':
+                if cls is SearchQuery:
                     # This check is needed because BulkSearchQuery inherits SearchQuery and because of the way we set
                     # query for it, it causes this check to fail since we previously provided queryTensor/queryLexical
                     # parameters
-                    if hybrid_parameters and (hybrid_parameters.queryTensor or hybrid_parameters.queryLexical):
+                    print("VALIDATING HERE")
+                    print(hybrid_parameters)
```
Comment (farshidz) [inline src/marqo/tensor_search/models/api_models.py]: Did you leave dev code behind?

File: src/marqo/tensor_search/models/search.py [hunk 17]
```diff
@@ -69,6 +69,7 @@ class SearchContext(BaseModel):
 
     def __init__(self, **data):
         try:
+            print(data)
```
Comment (farshidz) [inline src/marqo/tensor_search/models/search.py]: Remove

**Review by farshidz — CHANGES_REQUESTED** [review]

File: src/marqo/core/models/hybrid_parameters.py [hunk 18]
```diff
@@ -95,6 +95,22 @@ def validate_properties(cls, values):
             if values.get('rankingMethod') not in [RankingMethod.Lexical, RankingMethod.Tensor]:
                 raise ValueError("For retrievalMethod: tensor or lexical, rankingMethod must be: tensor or lexical")
 
+        # if one of queryLexical and queryTensor is present but other one is not
+        if (values.get('queryLexical') is None) != (values.get('queryTensor') is None):
+            if values.get('retrievalMethod') == RetrievalMethod.Disjunction:
+                raise ValueError(
+                    "Both queryLexical and queryTensor or q(Query) must be present when "
```
Comment (farshidz) [inline src/marqo/core/models/hybrid_parameters.py]: q(Query) seems confusing. You can just use q because it's an API parameter. They can look up the docs if needed

Also I think we do single quotes like `'queryLexical'` but check other errors to be sure

File: src/marqo/core/models/hybrid_parameters.py [hunk 19]
```diff
@@ -95,6 +95,22 @@ def validate_properties(cls, values):
...
+                    "disjunction retrieval method is used."
+                )
+
```
Comment (farshidz) [inline src/marqo/core/models/hybrid_parameters.py]: What about lexical/lexical, lexical/tensor, tensor/tensor?

In that oder, you expect
* Only a lexical query
* Both queries
* Only a tensor query

Or q in all cases

File: src/marqo/core/search/hybrid_search.py [hunk 20]
```diff
@@ -165,9 +189,16 @@ def search(
             hybridParameters=hybrid_parameters
         )]
 
-        with RequestMetricsStore.for_request().time(f"search.hybrid.vector_inference_full_pipeline"):
-            qidx_to_vectors: Dict[Qidx, List[float]] = run_vectorise_pipeline(config, queries, device)
-        vectorised_text = list(qidx_to_vectors.values())[0]
+        if not (
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py]: Why do you say not lexcial instead of saying it's tensor? It's confusing to read

File: tests/api_tests/v1/tests/api_tests/test_search_common.py [hunk 21]
```diff
@@ -222,18 +222,38 @@ def test_hybrid_search_validations(self):
                         )
                     assert "Query(q) cannot be provided for HYBRID search when hybridParameters.queryTensor or hybridParameters.queryLexical is provided" in str(e.exception)
 
+                with self.subTest("Hybrid search with only one queryTensor/queryLexical and retrievalMethod=disjunction raises an error"):
+                    with self.assertRaises(MarqoWebError) as e:
```
Comment (farshidz) [inline tests/api_tests/v1/tests/api_tests/test_search_common.py:230]: Message and code assertions missing, this could pass on any error

File: tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py [hunk 22]
```diff
@@ -3144,8 +3148,8 @@ def test_none_query_lexical(self):
                         config=self.config, index_name=index.name, search_method="HYBRID", text=None,
                         hybrid_parameters=HybridParameters(
                             queryLexical=None,
```
Comment (farshidz) [inline tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py:3109]: Should this not be an error though? What's the behavior? What are we searching?

File: tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py [hunk 23]
```diff
@@ -3214,3 +3218,49 @@ def test_none_query_tensor_disjunction_retrieval(self):
                             rankingMethod=RankingMethod.RRF
                         ), result_count=5
                     )
+
+    def test_none_provided_for_tensor_lexical_retrieval_works(self):
+        """Ensure that None can be provided for retrievalMethod and rankingMethod."""
```
Comment (farshidz) [inline tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py]: Description doesn't match test

File: tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py [hunk 24]
```diff
@@ -3214,3 +3218,49 @@ def test_none_query_tensor_disjunction_retrieval(self):
...
+                        queryLexical="dogs",
+                        retrievalMethod=RetrievalMethod.Lexical,
+                        rankingMethod=RankingMethod.Lexical,
+                    ), result_count=5
+                )
+
+                self.assertIn("hits", res)
+
+    def test_none_provided_for_lexical_tensor_retrieval_works(self):
+        """Ensure that None can be provided for retrievalMethod and rankingMethod."""
```
Comment (farshidz) [inline tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py]: Description doesn't match test

File: src/marqo/core/search/hybrid_search.py [hunk 25]
```diff
@@ -124,21 +124,45 @@ def search(
...
+            tensor_query = query.customVector.vector
+            lexical_query = query.customVector.content
+        else:
+            tensor_query = query
+            lexical_query = query
+
+        if (tensor_query is None) != (lexical_query is None):
+            if hybrid_parameters.retrievalMethod == RetrievalMethod.Disjunction:
+                raise core_exceptions.InvalidArgumentError(
+                    "Both queryLexical and queryTensor or q(Query) must be present when "
```
Comment (farshidz) [inline src/marqo/core/search/hybrid_search.py]: q(Query) seems confusing. You can just use q because it's an API parameter. They can look up the docs if needed

Also I think we do single quotes like 'queryLexical' but check other errors to be sure

File: tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py (hunk 22 above)
Comment (danyilq) [reply tests/integ_tests/tensor_search/integ_tests/test_hybrid_search.py:3109]: it is indeed an error, that's what being tested here, updated the description..

**Review by farshidz — DISMISSED** [review]
